server/session/handler_item_stack_request.go

```diff
@@ -550,8 +572,72 @@ func (h *ItemStackRequestHandler) reject(id int32, s *Session) {
 	h.changes = map[byte]map[byte]changeInfo{}
 }
 
-// hasRequiredInputs checks and validates the inputs for a crafting grid.
-func (h *ItemStackRequestHandler) hasRequiredInputs(inputs []recipe.Item, s *Session) bool {
+// inputData is used for input maps for inventory input checking and removing.
+type inputData struct {
+	// count is the count of the input.
+	count int
+	// metadataValue is the metadata value of the input.
+	metadataValue int16
+	// allTypes is true if the input applies to all of it's kind.
+	allTypes bool
+}
+
+// hasRequiredInventoryInputs checks and validates if the player inventory has the necessary inputs.
+func (h *ItemStackRequestHandler) hasRequiredInventoryInputs(inputs []recipe.Item, s *Session) bool {
+	inputMap := make(map[string]inputData)
+	for _, input := range inputs {
+		it := input.Item()
+		if it == nil {
+			continue
+		}
+
+		name, meta := it.EncodeItem()
+		data := inputData{
+			count:         input.Count(),
+			metadataValue: meta,
+			allTypes:      input.AllTypes,
+		}
+
+		if newData, ok := inputMap[name]; ok {
+			if newData.metadataValue == data.metadataValue || newData.allTypes && data.allTypes {
+				data.count = data.count + newData.count
+			}
+		}
+
+		inputMap[name] = data
+	}
+
+	for slot := 0; slot < s.inv.Size(); slot++ {
```

Comment:
I would use `(*Inventory).Contents()` here

Reply:
Contents doesn't include the slot which is needed, and I can't just assume the slot because it doesn't include empty stacks.

Reply:
Changed.